File: tvr-decoder/src/main/java/com/willcurrie/decoders/CVMResultsDecoder.java
package com.willcurrie.decoders;

import java.util.Arrays;
import java.util.List;

import com.willcurrie.DecodedData;
import com.willcurrie.Decoder;
import com.willcurrie.tlv.ISOUtil;

public class CVMResultsDecoder implements Decoder {

	private static final int FIELD_LENGTH = 6;

	@Override
	public List<DecodedData> decode(String input, int startIndexInBytes) {
		CVRule rule = new CVRule(input.substring(0, 4));
		String result = input.substring(4, 6);
		return Arrays.asList(
				new DecodedData(input.substring(0, 2), rule.getVerificationMethod().getDescription(), startIndexInBytes, startIndexInBytes + 1),
				new DecodedData(input.substring(2, 4), rule.getConditionCode().getDescription(), startIndexInBytes + 1, startIndexInBytes + 2),
				new DecodedData(result, decodeResult(result), startIndexInBytes + 2, startIndexInBytes + 3)
				);
	}

	private String decodeResult(String result) {
		return "01".equals(result) ? "Failed" :
			   "02".equals(result) ? "Sucessful" : "Unknown";
	}
	
	@Override
	public int getMaxLength() {
		return FIELD_LENGTH;
	}

	@Override
	public String validate(String bitString) {
		if (bitString == null || bitString.length() != FIELD_LENGTH) {
			return String.format("Value must be exactly %d characters", FIELD_LENGTH); 
		}
		if (!ISOUtil.isValidHexString(bitString)) {
			return "Value must contain only the characters 0-9 and A-F";
		}
		return null;
	}

}

File: src/main/java/io/github/binaryfoo/decoders/CVMResultsDecoder.kt
package io.github.binaryfoo.decoders

import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.Decoder
import io.github.binaryfoo.tlv.ISOUtil

class CVMResultsDecoder : Decoder {

  override fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): List<DecodedData> {
    val rule = CVRule(input.substring(0, 4))
    val result = input.substring(4, 6)
    return listOf(
        DecodedData.primitive(input.substring(0, 2), rule.verificationMethodDescription, startIndexInBytes, startIndexInBytes + 1),
        DecodedData.primitive(input.substring(2, 4), rule.conditionCodeDescription, startIndexInBytes + 1, startIndexInBytes + 2),
        DecodedData.primitive(result, decodeResult(result), startIndexInBytes + 2, startIndexInBytes + 3))
  }

  private fun decodeResult(result: String): String {
    return if ("01" == result)
      "Failed"
    else if ("02" == result) "Sucessful" else "Unknown"
  }

  override fun getMaxLength(): Int = FIELD_LENGTH

  override fun validate(input: String?): String? {
    if (input == null || input.length != FIELD_LENGTH) {
      return "Value must be exactly $FIELD_LENGTH characters"
    }
    if (!ISOUtil.isValidHexString(input)) {
      return "Value must contain only the characters 0-9 and A-F"
    }
    return null
  }

  companion object {
    private val FIELD_LENGTH = 6
  }

}
